Ability to set a Global Volume

LiamGallagher737 opened the pull request:
# Objective

Adds a new resource to control a global volume.
Fixes #7690

---

## Solution

Added a new resource to control global volume, this is then multiplied with an audio sources volume to get the output volume, individual audio sources can opt out of this my enabling the `absolute_volume` field in `PlaybackSettings`.

---

## Changelog

### Added
- `GlobalVolume` a resource to control global volume (in prelude).
- `global_volume` field to `AudioPlugin` or setting the initial value of `GlobalVolume`.
- `Volume` enum that can be `Relative` or `Absolute`.
- `VolumeLevel` struct for defining a volume level.

LiamGallagher737 commented:
This also opens the opportunity to fix the issue in the `decodable` audio example being very loud by adding `.insert_resource(GlobalVolume::new(0.2))` to it. Tested it and it works.

esensar commented on crates/bevy_audio/src/audio_output.rs:
```diff
@@ -79,7 +80,7 @@ where
                 if let Some(sink) = self.play_source(audio_source, config.settings.repeat) {
                     sink.set_speed(config.settings.speed);
-                    sink.set_volume(config.settings.volume);
+                    sink.set_volume(config.settings.volume * global_volume.volume);
```
Should the `absolute_volume` flag be used here?

LiamGallagher737 replied:
Yeah, don't recommend submitting code late at night

esensar commented on crates/bevy_audio/src/audio.rs:
```diff
@@ -128,6 +128,8 @@ pub struct PlaybackSettings {
     /// Volume to play at.
     pub volume: f32,
+    /// Ignore global volume
+    pub absolue_volume: bool,
```
Typo here

alice-i-cecile replied:
```suggestion
    pub absolute_volume: bool,
```

This will need a rename elsewhere though.

esensar commented on crates/bevy_audio/src/audio.rs:
```diff
@@ -164,6 +168,12 @@ impl PlaybackSettings {
     }
+
+    /// Helper to set if the volume sould ignore the global volume.
+    pub const fn with_absolute_volume(mut self, absolue_volume: bool) -> Self {
```
Typo here too

st0rmbtw commented on crates/bevy_audio/src/audio.rs:
```diff
@@ -128,6 +128,8 @@ pub struct PlaybackSettings {
     /// Volume to play at.
     pub volume: f32,
+    /// Ignore global volume
+    pub absolue_volume: bool,
```
A small typo here

esensar commented on crates/bevy_audio/src/audio.rs:
```diff
@@ -164,6 +168,12 @@ impl PlaybackSettings {
         self
     }
+
+    /// Helper to set if the volume sould ignore the global volume.
```
Just noticed the typo in the comment too (should)

alice-i-cecile commented on crates/bevy_audio/src/audio.rs:
```diff
@@ -188,3 +198,23 @@ where
     }
 }
+
+/// Use this [`Resource`] to control the global volume of all non-absolute audio.
```
This could be a bit clearer about what "non-absolute" means :)

LiamGallagher737 commented on crates/bevy_audio/src/audio_output.rs:
```diff
@@ -79,7 +80,11 @@ where
                 if let Some(sink) = self.play_source(audio_source, config.settings.repeat) {
                     sink.set_speed(config.settings.speed);
-                    sink.set_volume(config.settings.volume);
+                    if config.settings.absolute_volume {
```
Could be replaced with something like this if it's preferred.
```rs
sink.set_volume(config.settings.volume * if config.settings.absolute_volume { 1.0 } else { global_volume.volume });
```

alice-i-cecile replied:
I think this is a bit more clear.

LiamGallagher737 replied:
The current solution or the proposed one?

alice-i-cecile replied:
Current solution.

alice-i-cecile commented on crates/bevy_audio/src/audio.rs:
```diff
@@ -188,3 +198,24 @@ where
 }
+
+/// Use this [`Resource`] to control the global volume of all non-absolute audio.
+/// Non-abosolute being any audio source that doesn't have [`absolute_volume`](PlaybackSettings::absolute_volume) set to `true`.
```
```suggestion
/// Sources can be set to ignore this setting by setting the `absolute_volume` field on [`PlaybackSettings`] to `true`.
```

alice-i-cecile replied:
I think this is clearer.

alice-i-cecile reviewed (approved):
Clarity nit in the docs, but I won't block on it.

mockersf reviewed (changes requested):
Changing the global volume should change the sounds already playing that have been set relative to it.

Or as a start it should be mentioned that setting the global volume will only impact new sounds played after but not sounds already playing.

I think I would prefer an enum for the volume instead of having two fields, but I'm not completely sure on that.

alice-i-cecile commented:
> Changing the global volume should change the sounds already playing that have been set relative to it.

Agreed, but this may not be trivial. Docs would be fine for me as a start.

> I think I would prefer an enum for the volume instead of having two fields, but I'm not completely sure on that.

This is nice :)

LiamGallagher737 commented:
Enum for volume makes a lot of sense, I'll have a go at changing the volume of already playing sounds.

alice-i-cecile commented:
That looks perfect. Ideally we could document what the `f32` means, or use a more sensible numeric type too. Negative volumes don't exactly make sense to me.

LiamGallagher737 commented:
Is there a way in rust to have positive only floats?

alice-i-cecile commented:
> Is there a way in rust to have positive only floats?

I would newtype here and use a constructor method.

LiamGallagher737 commented:
This is what I have got so far
```rs
/// Defines the volume to play an audio source at.
pub enum Volume {
    /// A volume level relative to the global volume.
    Relative(VolumeLevel),
    /// A volume level that ignores the global volume.
    Absolute(VolumeLevel),
}

impl Default for Volume {
    fn default() -> Self {
        Self::Relative(VolumeLevel::default())
    }
}

/// not sure what to put here
pub struct VolumeLevel(f32);

impl Default for VolumeLevel {
    fn default() -> Self {
        Self(1.0)
    }
}

impl VolumeLevel {
    /// Create a new volume level.
    pub fn new(volume: f32) -> Self {
        debug_assert!(volume >= 0.0);
        Self(volume)
    }
}
```

But if feels very verbose to me, personally I think it's fine using an `f32`. Some thoughts on this would be great.

LiamGallagher737 commented:
I've added the volume enum but using the newtype for the volume level feels too verbose imo. I also added the global volume as a field in the audio struct so you can set it using the `.set()` method on the plugin group.

LiamGallagher737 commented:
> Changing the global volume should change the sounds already playing that have been set relative to it.

Updating the volume for AudioSinks that have not been detached is easy, but I am unsure of how to update ones which have been detached (have no strong handle) so for now I will add message mentioning that only new sounds will be impacted.

esensar reviewed (approved):
Nice!
It is a bit more verbose, but it prevent missing the absolute option completely.

MinerSebas commented on crates/bevy_audio/src/lib.rs:
```diff
@@ -48,14 +48,18 @@ use bevy_ecs::prelude::*;
 #[derive(Default)]
-pub struct AudioPlugin;
+pub struct AudioPlugin {
+    /// The global volume for all audio sources with a [`Volume::Relative] volume.
```
```suggestion
    /// The global volume for all audio sources with a [`Volume::Relative`] volume.
```

MinerSebas commented on crates/bevy_audio/src/audio.rs:
```diff
@@ -121,13 +121,57 @@ where
+
+/// A volume level equivalent to a positive only float.
+#[derive(Clone, Copy, Debug)]
+pub struct VolumeLevel(pub(crate) f32);
```
This struct is missing a Getter for the inner value.

MinerSebas commented on crates/bevy_audio/src/audio.rs:
```diff
@@ -188,3 +232,22 @@ where
 }
+
+/// Use this [`Resource`] to control the global volume of all non-absolute audio.
+/// Sources can be set to ignore this setting by setting the `absolute_volume` field on [`PlaybackSettings`] to `true`.
```
This sentence is wrong, as that change is done using the `Volume` enum and not a bool.

mockersf commented on crates/bevy_audio/src/audio.rs:
```diff
@@ -121,13 +121,61 @@ where
+
+/// A volume level equivalent to a positive only float.
+#[derive(Clone, Copy, Debug)]
+pub struct VolumeLevel(pub(crate) f32);
```
You could also implements `Deref` for this struct

LiamGallagher737 commented:
> Ideally we could document what the f32 means, or use a more sensible numeric type too. Negative volumes don't exactly make sense to me.

Still not quite happy with the use of `VolumeLevel`
It turns
```rust
.with_volume(0.75)
```
into
```rust
.with_volume(Volume::Relative(VolumeLevel::new(0.75)))
```
I personally think just using a f32 like this is fine, I think it's unlikely someone will give a negative value
```rust
.with_volume(Volume::Relative(0.75))
```
I'm going to remove `VolumeLevel` and and a doc letting people know it must be positive instead, lmk if you want me to bring `VolumeLevel` back

alice-i-cecile commented:
I'm glad to see this added. Merging now, we can refine later as needed.

mockersf commented on examples/audio/decodable.rs:
```diff
@@ -1,6 +1,7 @@
 //! Shows how to create a custom `Decodable` type by implementing a Sine wave.
 //! ***WARNING THIS EXAMPLE IS VERY LOUD.*** Turn your volume down.
```
is this comment still true?

mockersf commented on crates/bevy_audio/src/audio.rs:
```diff
@@ -251,13 +255,75 @@ where
+    }
+}
+
+impl Deref for VolumeLevel {
+    type Target = f32;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for VolumeLevel {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
```
you could derive `Deref` and `DerefMut`
